File: 20-valid-parentheses/solution.ts
function isValid(s: string): boolean {
  const stack: string[] = [];
  const length = s.length;

  for (let i = 0; i < length; i++) {
    const c = s[i];
    if (['(', '[', '{'].includes(c)) {
      stack.push(c);
    } else {
      const last = stack.pop();
      if (
        (c === ')' && last !== '(') ||
        (c === ']' && last !== '[') ||
        (c === '}' && last !== '{')
      ) {
        return false;
      }
    }
  }
  return stack.length === 0;
}

File: 20-valid-parentheses/solution.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> sk;
        for(auto x : s) {
            if(x == '(' || x == '{' || x == '[') sk.push(x);
            else if(x == ')') {
                if(sk.empty() || sk.top() != '(') return false;
                sk.pop();
            } else if(x == '}') {
                if(sk.empty() ||sk.top() != '{') return false;
                sk.pop();
            } else {
                if(sk.empty() ||sk.top() != '[') return false;
                sk.pop();
            }
        }
        return sk.empty();
    }
};
